analyses/cms-open-data-ttbar/analysis.py

```diff
@@ -1,518 +1,361 @@
-PARSER.add_argument("-v", "--verbose", action="store_true")
-ARGS = PARSER.parse_args()
-
-# importing ROOT takes a little while, so we only do it if argument parsing succeeded
-import ROOT # noqa: {E402}
-
-if ARGS.scheduling_mode == "imt":
-
-    RDataFrame = ROOT.RDataFrame
-    RunGraphs = ROOT.RDF.RunGraphs
-    VariationsFor = ROOT.RDF.Experimental.VariationsFor
-
-    def init_functions():
-        ROOT.gSystem.CompileMacro("helper.cpp", "kO")
-
-else:
-
-    # Dask configuration useful in distributed mode
-    RDataFrame = ROOT.RDF.Experimental.Distributed.Dask.RDataFrame
-    RunGraphs = ROOT.RDF.Experimental.Distributed.RunGraphs
-    VariationsFor = ROOT.RDF.Experimental.Distributed.VariationsFor
-    initialize = ROOT.RDF.Experimental.Distributed.initialize
-
-    if ARGS.scheduling_mode == "dask-ssh":
-        if not ARGS.nodes:
-            raise ValueError("For SSHCluster deployments, please specify a "
-                             "string with a comma-separated list of hostnames of nodes that will be used.")
-
-        if ARGS.npartitions is None:
-            n_compute_nodes = len(ARGS.nodes.split(",")) - 1
-            ARGS.npartitions = ARGS.ncores * n_compute_nodes
-
-    elif ARGS.scheduling_mode == "dask-local":
-        if ARGS.npartitions is None:
-            ARGS.npartitions = ARGS.ncores
-
-    from distributed import Client, LocalCluster, SSHCluster, get_worker
-
-    def create_localcluster_connection(ncores: int) -> Client:
-        cluster = LocalCluster(n_workers=ncores, threads_per_worker=1, processes=True)
-        client = Client(cluster)
-        return client
-
-    def create_sshcluster_connection(nodes: str, ncores: int) -> Client:
-        parsed_nodes = nodes.split(',')
-        scheduler = parsed_nodes[:1]
-        workers = parsed_nodes[1:]
-
-        print(f"List of nodes: {scheduler=}, {workers=}")
-
-        # The creation of the SSHCluster object needs to be further configured according to needs.
+from time import time
+from typing import Optional
+
+import ROOT
+from distributed import Client, LocalCluster, SSHCluster, get_worker
+from plotting import save_plots
+from utils import (
+    AGCInput,
+    AGCResult,
+    postprocess_results,
+    retrieve_inputs,
+    save_histos,
+)
+
+# Using https://atlas-groupdata.web.cern.ch/atlas-groupdata/dev/AnalysisTop/TopDataPreparation/XSection-MC15-13TeV.data
+# as a reference. Values are in pb.
+XSEC_INFO = {
+    "ttbar": 396.87 + 332.97,  # nonallhad + allhad, keep same x-sec for all
+    "single_top_s_chan": 2.0268 + 1.2676,
+    "single_top_t_chan": (36.993 + 22.175) / 0.252,  # scale from lepton filter to inclusive
+    "single_top_tW": 37.936 + 37.906,
+    "wjets": 61457 * 0.252,  # e/mu+nu final states
+}
+
+
+def parse_args() -> argparse.Namespace:
```

Comment:
Moving the argument parsing logic to a function means that we now have to wait for all the `import` statements (including the slightly costly `import ROOT`) to end before we could even print the help, e.g.
```
python analysis.py -h
```

I think it would make for a smoother experience if this logic was moved to before `import ROOT`

Reply:
we can do that, but we have to explain to the ruff linter and to readers why we do that. I didn't want to burden the code with that, to just focus on RDF things (plus I find it a bit awkward that we even have to think about that when importing ROOT). if you are running the AGC script you don't expect it to run in less than a second anyway.

if you feel strongly about this I can move the `import ROOT` down, silence the linting error and add an explanation in a comment.

Reply:
That's a fair point, let's leave it as it is